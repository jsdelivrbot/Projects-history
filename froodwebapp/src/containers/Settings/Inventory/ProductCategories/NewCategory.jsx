import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Row, Col } from 'react-flexbox-grid';
import { FormInput } from 'components';
import { reduxForm, Field } from 'redux-form';
import { Button } from 'antd';
import {
  prodCatSaveRequest
} from 'redux-base/actions';
import { form } from 'styles/common.scss';

const mapDispatchToProps = {
  prodCatSaveRequest
};

const reduxFormConfig = {
  form: 'newProductCategoryForm',
};

export class NewCategory extends Component {
  handleSave = (values) => {
    this.props.prodCatSaveRequest({
      payload: {
        name: values.name,
        parentId: 0
      }
    });
  }

  render() {
    const { handleSubmit, pristine } = this.props;

    return (
      <form
        className={ form }
        onSubmit={ handleSubmit(this.handleSave) }
      >
        <Row middle="xs" center="xs">
          <Col>
            <h3>No product categories found</h3>
          </Col>
        </Row>
        <Row middle="xs" center="xs">
          <Col xs={ 12 } sm={ 12 } md={ 4 } lg={ 4 }>
            <Field
              name="name"
              placeholder="Add your first parent category"
              component={ FormInput }
            />
          </Col>
        </Row>
        <Row middle="xs" center="xs">
          <Col>
            <Button
              type="primary"
              htmlType="submit"
              disabled={ pristine }
            >
              Add
            </Button>
          </Col>
        </Row>
      </form>
    );
  }
}

NewCategory.propTypes = {
  // form triggers
  pristine: PropTypes.bool.isRequired,
  // redux-base
  prodCatSaveRequest: PropTypes.func.isRequired,
  // form handleSubmit
  handleSubmit: PropTypes.func.isRequired
};

export default connect(null, mapDispatchToProps)(reduxForm(reduxFormConfig)(NewCategory));
