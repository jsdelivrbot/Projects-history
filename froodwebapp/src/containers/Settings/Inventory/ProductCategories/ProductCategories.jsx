/* eslint-disable jsx-a11y/label-has-for */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { prodCatGetRequest } from 'redux-base/actions';
import { connect } from 'react-redux';
import { Spin } from 'antd';
import { PageHeader } from 'components';
import ConnectedProductsTree from './ProductsTree';
import ConnectedNewCategory from './NewCategory';

const mapStateToProps = state => ({
  loadingPage: state.productCategories.loadingPage,
  needReloadProductCategories: state.productCategories.needReloadProductCategories,
  productCategories: state.productCategories.data
});

const mapDispatchToProps = {
  prodCatGetRequest,
};

export class ProductCategories extends Component {
  componentWillMount = () => {
    this.props.prodCatGetRequest();
  }

  componentWillReceiveProps(nextProps) {
    // when we save, update or delete category
    if (nextProps.needReloadProductCategories) {
      this.props.prodCatGetRequest();
    }
  }

  render() {
    const {
      loadingPage,
      productCategories
    } = this.props;

    return (
      <div>
        <PageHeader
          bigText="Product categories"
          smallText="Manage categories hierarchy"
        />
        <Spin spinning={ loadingPage }>
          { (productCategories.length === 0 && !loadingPage) && <ConnectedNewCategory /> }
          { (productCategories.length > 0) && <ConnectedProductsTree /> }
        </Spin>
      </div>
    );
  }
}

ProductCategories.propTypes = {
  // triggers
  loadingPage: PropTypes.bool.isRequired,
  needReloadProductCategories: PropTypes.bool.isRequired,
  // data
  productCategories: PropTypes.array.isRequired,
  // redux-base
  prodCatGetRequest: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(ProductCategories);
