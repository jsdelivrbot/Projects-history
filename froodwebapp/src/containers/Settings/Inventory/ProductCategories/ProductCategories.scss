@import '~styles/colors.scss';

.highlight {
  color: $main-red;
}

.treeNode {
  &::before {
    margin: 28px 0 !important;
  }
  span[class*='ant-tree-node-content-wrapper'] {
    vertical-align: baseline;
  }
}


