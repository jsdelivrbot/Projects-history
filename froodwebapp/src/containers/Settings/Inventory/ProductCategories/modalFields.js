export default () => ([{
  header: 'Name',
  type: 'text',
  name: 'name',
  editableForExistingItem: true,
}]);
