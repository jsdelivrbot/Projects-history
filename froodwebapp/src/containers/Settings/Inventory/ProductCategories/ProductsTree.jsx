import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  prodCatDeleteRequest,
  prodCatSaveRequest
} from 'redux-base/actions';
import { Row, Col } from 'react-flexbox-grid';
import { connect } from 'react-redux';
import { ConnectedTopFormModal } from 'components';
import { Tree, Dropdown, Button, Icon, Menu, Input } from 'antd';
import classnames from 'classnames';
import fields from './modalFields';
import { highlight, treeNode } from './ProductCategories.scss';

const { TreeNode } = Tree;
const { Search } = Input;

const mapStateToProps = state => ({
  loadingPage: state.productCategories.loadingPage,
  needReloadProductCategories: state.productCategories.needReloadProductCategories,
  productCategories: state.productCategories.data
});

const mapDispatchToProps = {
  prodCatDeleteRequest,
  prodCatSaveRequest
};

const newCategory = {
  name: ''
};

export class ProductsTree extends Component {
  constructor(props) {
    super(props);
    this.state = {
      modalVisible: false,
      modalData: newCategory,
      searchText: '',
      autoExpandParent: true,
      expandedKeys: [],
      nodesList: []
    };
  }


  componentWillMount() {
    const nodesList = [];
    this.getNodesList(this.props.productCategories, nodesList);

    this.setState({
      nodesList
    });
  }

  componentWillReceiveProps(nextProps) {
    const nodesList = [];
    this.getNodesList(nextProps.productCategories, nodesList);

    if (nextProps.needReloadProductCategories) {
      this.setState({
        modalVisible: false,
        modalData: undefined
      });
    }

    if (this.state.nodesList.length !== nodesList.length) {
      this.setState({
        modalVisible: false,
        modalData: {},
        nodesList
      });
    }
  }

  getNodesList = (nodes, result) => {
    nodes.forEach((node) => {
      const { name, children } = node;

      result.push({ name, title: name });

      if (children) {
        this.getNodesList(children, result);
      }
    });
  }

  // helpers
  getParentKey = (nodeName, nodes) => {
    let parentKey;
    if (nodes) {
      nodes.forEach((node) => {
        if (node.children && node.children.some(cnode => cnode.name === nodeName)) {
          parentKey = node.name;
        } else if (this.getParentKey(nodeName, node.children)) {
          parentKey = this.getParentKey(nodeName, node.children);
        }
      });
    }
    return parentKey;
  };

  getMenu = ({
    id,
    parentId,
    showAddSibling = false
  }) => (
    <Menu onClick={ this.handleMenuClick }>
      { showAddSibling &&
        <Menu.Item activeKey={ { action: 'addSibling', parentId } }>Add Sibling</Menu.Item>
      }
      <Menu.Item activeKey={ { action: 'addChild', id } }>Add Child</Menu.Item>
      <Menu.Item activeKey={ { action: 'delete', id } }>Delete</Menu.Item>
    </Menu>
  );

  getDropdown = (title, menu, highlightText = false) => (
    <Dropdown
      overlay={ menu }
    >
      <Button className={ classnames({ [highlight]: highlightText }) }>
        { title } <Icon type="down" />
      </Button>
    </Dropdown>
  );

  getTreeNodes = (nodes, parentId, searchText) => (
    nodes.map((node) => {
      // Matches searched text with categories to highlight them.
      const highlightText = searchText && node.name.toLowerCase().includes(searchText.toLowerCase());

      if (node.children) {
        return (
          <TreeNode
            className={ treeNode }
            key={ node.name }
            title={
              this.getDropdown(
                node.name,
                this.getMenu({
                  id: node.id,
                  parentId,
                  showAddSibling: true
                }),
                highlightText
              )
            }
          >
            { this.getTreeNodes(node.children, node.id, searchText) }
          </TreeNode>
        );
      }
      return (
        <TreeNode
          className={ treeNode }
          key={ node.name }
          title={
            this.getDropdown(
              node.name,
              this.getMenu({
                id: node.id,
                parentId,
                showAddSibling: true
              }),
              highlightText
            )
          }
        />
      );
    })
  )

  handleToggleModal = () => {
    this.setState({
      modalVisible: !this.state.modalVisible,
      modalData: newCategory,
    });
  }

  handleSearchChange = (e) => {
    const searchText = e.target.value.toLowerCase();

    // get list of tree nodes to expand.
    const expandedKeys = [];
    this.state.nodesList.forEach((node) => {
      if (node.name.toLowerCase().includes(searchText.toLowerCase())) {
        const expandedKey = this.getParentKey(node.name, this.props.productCategories);
        if (expandedKey) expandedKeys.push(expandedKey);
      }
    });

    this.setState({
      expandedKeys,
      searchText,
      autoExpandParent: true
    });
  }

  handleTreeNodeExpand = (expandedKeys) => {
    this.setState({
      expandedKeys,
      autoExpandParent: false,
    });
  }

  handleMenuClick = ({ item: { props: { activeKey } } }) => {
    switch (activeKey.action) {
      case 'addSibling':
        this.handleSetParentIdForModalData(activeKey.parentId);
        break;
      case 'addChild':
        this.handleSetParentIdForModalData(activeKey.id);
        break;
      case 'delete':
        this.handleDeleteCategory(activeKey.id);
        break;
      default:
        break;
    }
  }

  handleSetParentIdForModalData = (id) => {
    this.setState({
      modalVisible: true,
      modalData: {
        parentId: id,
        name: ''
      }
    });
  }

  handleDeleteCategory = (id) => {
    this.props.prodCatDeleteRequest({ id });
  }

  handleSaveCaterogy = (values) => {
    this.props.prodCatSaveRequest({
      payload: values
    });
  }

  render() {
    const {
      modalData,
      modalVisible,
      expandedKeys,
      searchText,
      autoExpandParent
    } = this.state;

    const {
      loadingPage,
      productCategories
    } = this.props;

    return (
      <div>
        <ConnectedTopFormModal
          loading={ loadingPage }
          title="New Category"
          buttonVisible={ false }
          handleToggleModal={ this.handleToggleModal }
          handleSave={ this.handleSaveCaterogy }
          visible={ modalVisible }
          initialValues={ modalData }
          fields={ fields() }
        />
        <Row>
          <Col xs={ 3 }>
            <Search
              placeholder="Search"
              onChange={ this.handleSearchChange }
            />
          </Col>
        </Row>
        <Row>
          <Col>
            <Tree
              showLine
              onExpand={ this.handleTreeNodeExpand }
              expandedKeys={ expandedKeys }
              autoExpandParent={ autoExpandParent }
            >
              { this.getTreeNodes(productCategories, 0, searchText) }
            </Tree>
          </Col>
        </Row>
      </div>
    );
  }
}

ProductsTree.propTypes = {
  // triggers
  loadingPage: PropTypes.bool.isRequired,
  needReloadProductCategories: PropTypes.bool.isRequired,
  // data
  productCategories: PropTypes.array.isRequired,
  // redux-base
  prodCatDeleteRequest: PropTypes.func.isRequired,
  prodCatSaveRequest: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(ProductsTree);
