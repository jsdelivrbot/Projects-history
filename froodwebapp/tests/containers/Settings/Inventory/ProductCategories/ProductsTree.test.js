import { ProductsTree } from 'containers/Settings/Inventory/ProductCategories/ProductsTree';
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const { describe, expect, it } = global;

describe('ProductsTree', () => {
  const props = {
    loadingPage: false,
    needReloadProductCategories: false,
    productCategories: [],
    prodCatGetRequest: jest.fn(),
    prodCatDeleteRequest: jest.fn(),
    prodCatSaveRequest: jest.fn()
  };

  it('renders the ProductsTree component', () => {
    const component = shallow(<ProductsTree { ...props } />);

    expect(component.length).toEqual(1);
  });

  it('capturing Snapshot of ProductsTree', () => {
    const component = shallow(<ProductsTree { ...props } />);
    const view = toJson(component);
    const actual = expect(view);

    actual.toMatchSnapshot();
  });
});
