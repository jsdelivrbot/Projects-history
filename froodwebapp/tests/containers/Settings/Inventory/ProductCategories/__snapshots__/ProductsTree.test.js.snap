// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsTree capturing Snapshot of ProductsTree 1`] = `
<div>
  <ReduxForm
    buttonVisible={false}
    fields={
      Array [
        Object {
          "editableForExistingItem": true,
          "header": "Name",
          "name": "name",
          "type": "text",
        },
      ]
    }
    handleSave={[Function]}
    handleToggleModal={[Function]}
    initialValues={
      Object {
        "name": "",
      }
    }
    loading={false}
    title="New Category"
    visible={false}
  />
  <Row>
    <Col
      xs={3}
    >
      <Search
        onChange={[Function]}
        placeholder="Search"
        prefixCls="ant-input-search"
      />
    </Col>
  </Row>
  <Row>
    <Col>
      <Tree
        autoExpandParent={true}
        checkable={false}
        expandedKeys={Array []}
        onExpand={[Function]}
        openAnimation={
          Object {
            "appear": [Function],
            "enter": [Function],
            "leave": [Function],
          }
        }
        prefixCls="ant-tree"
        showIcon={false}
        showLine={true}
      >
        <TreeNode
          className="treeNode"
          key="products"
          title={
            <Dropdown
              mouseEnterDelay={0.15}
              mouseLeaveDelay={0.1}
              overlay={
                <Menu
                  className=""
                  onClick={[Function]}
                  prefixCls="ant-menu"
                  theme="light"
                >
                  <MenuItem
                    activeKey={
                      Object {
                        "action": "addChild",
                        "id": 0,
                      }
                    }
                  >
                    Add Child
                  </MenuItem>
                  <MenuItem
                    activeKey={
                      Object {
                        "action": "delete",
                        "id": 0,
                      }
                    }
                  >
                    Delete
                  </MenuItem>
                </Menu>
              }
              placement="bottomLeft"
              prefixCls="ant-dropdown"
            >
              <Button
                clicked={false}
                ghost={false}
                loading={false}
                prefixCls="ant-btn"
              >
                All Products 
                <Icon
                  type="down"
                />
              </Button>
            </Dropdown>
          }
        />
      </Tree>
    </Col>
  </Row>
</div>
`;
